transport/testing/reqres.go

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package test_transport
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"testing"
+
+	"github.com/yarpc/yarpc-go/transport"
+)
+
+// RequestMatcher may be used in gomock argument lists to assert that two
+// requests match.
+//
+// Requests are considered to be matching if:
+//
+// - All their primitive parameters match.
+// - The headers of the received request include all the headers from the
+//   source request. It may include extra headers.
+// - The contents of the request bodies are the same.
+type RequestMatcher struct {
+	t    *testing.T
+	req  *transport.Request
+	body []byte
+}
+
+// NewRequestMatcher constructs a new RequestMatcher from the given testing.T
+// and request.
+//
+// The request's contents are read in their entirety and replaced with a
+// bytes.Reader.
+func NewRequestMatcher(t *testing.T, r *transport.Request) RequestMatcher {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		t.Fatalf("failed to read request body: %v", err)
+		panic(err.Error())
+	}
+
+	// restore a copy of the body so that the caller can still use the request
+	// object
+	r.Body = bytes.NewReader(body)
+	return RequestMatcher{t: t, req: r, body: body}
+}
+
+// TODO: Headers like User-Agent, Content-Length, etc. make their way to the
+// user-level Request object. For now, we're doing the super set check but we
+// should do something more specific once yarpc/yarpc.io#2 is resolved.
+
+// Matches checks if the given object matches the Request provided in
+// NewRequestMatcher.
+func (m RequestMatcher) Matches(got interface{}) bool {
+	l := m.req
+	r, ok := got.(*transport.Request)
+	if !ok {
+		panic(fmt.Sprintf("expected *transport.Request, got %v", got))
+	}
+
+	if l.Caller != r.Caller {
```

Comment:
I see what you mean by can't just use `reflect.DeepEqual`, this is painful. I only have very hacky ideas,
e.g.
```go
// read r.Body into rBody
r.Body = nil
m.Body = nil
sameFields := r == m
r.Body = rBody
m.Body = mBody
```

since the `Body` field is being mutated anyway, technically this should be fine in code that doesn't have data races.. but it also feels wrong.